# Lab | Hypothesis Testing

**Objective**

Welcome to the Hypothesis Testing Lab, where we embark on an enlightening journey through the realm of statistical decision-making! In this laboratory, we delve into various scenarios, applying the powerful tools of hypothesis testing to scrutinize and interpret data.

From testing the mean of a single sample (One Sample T-Test), to investigating differences between independent groups (Two Sample T-Test), and exploring relationships within dependent samples (Paired Sample T-Test), our exploration knows no bounds. Furthermore, we'll venture into the realm of Analysis of Variance (ANOVA), unraveling the complexities of comparing means across multiple groups.

So, grab your statistical tools, prepare your hypotheses, and let's embark on this fascinating journey of exploration and discovery in the world of hypothesis testing!

**Challenge 1**

In this challenge, we will be working with pokemon data. The data can be found here:

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv

In [14]:
#libraries
import pandas as pd
import scipy.stats as st
import numpy as np
from scipy.stats import ttest_ind



In [3]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv")
df

Name   Type 1  Type 2  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0         Bulbasaur    Grass  Poison  45      49       49       65       65   
1           Ivysaur    Grass  Poison  60      62       63       80       80   
2          Venusaur    Grass  Poison  80      82       83      100      100   
3     Mega Venusaur    Grass  Poison  80     100      123      122      120   
4        Charmander     Fire     NaN  39      52       43       60       50   
..              ...      ...     ...  ..     ...      ...      ...      ...   
795         Diancie     Rock   Fairy  50     100      150      100      150   
796    Mega Diancie     Rock   Fairy  50     160      110      160      110   
797  Hoopa Confined  Psychic   Ghost  80     110       60      150      130   
798   Hoopa Unbound  Psychic    Dark  80     160       60      170      130   
799       Volcanion     Fire   Water  80     110      120      130       90   

     Speed  Generation  Legendary  
0       45           1      False  
1       60           1      False  
2       80           1      False  
3       80           1      False  
4       65           1      False  
..     ...         ...        ...  
795     50           6       True  
796    110           6       True  
797     70           6       True  
798     80           6       True  
799     70           6       True  

[800 rows x 11 columns]

- We posit that Pokemons of type Dragon have, on average, more HP stats than Grass. Choose the propper test and, with 5% significance, comment your findings.

__Hypothesis:__
* Null Hypothesis: There is no significant difference in the average HP between _Dragon_ type and _Grass_ type Pokémon.
* Alternative Hypothesis: _Dragon_ type Pokémon have, on average, higher HP than  _Grass_ type Pokémon.

In [5]:
# Filter Dragon-type and Grass-type Pokémon
dragon_hp = df[df['Type 1'] == 'Dragon']['HP']
grass_hp = df[df['Type 1'] == 'Grass']['HP']

In [9]:
# Perform a t-test (equal_var=False allows for unequal variance between the groups)
t_stat, p_value = ttest_ind(dragon_hp, grass_hp, equal_var=False)

# For one-tailed test, divide p-value by 2
p_value_one_tailed = p_value / 2

# Output the results
print(f"T-statistic: {t_stat:.4f}")
print(f"One-tailed p-value: {p_value_one_tailed:.4f}")

T-statistic: 3.3350
One-tailed p-value: 0.0008


__Conclusion:__
- The one-tailed p-value is 0.0008, which is much smaller than the 5% significance level (0.05).
- Since 𝑝 < 0.05, we reject the null hypothesis.

<br>
<br>

- We posit that Legendary Pokemons have different stats (HP, Attack, Defense, Sp.Atk, Sp.Def, Speed) when comparing with Non-Legendary. Choose the propper test and, with 5% significance, comment your findings.


__Hypothesis for Each Stat:__

For each stat (HP, Attack, Defense, Sp. Atk, Sp. Def, Speed), we will set up the following hypotheses:
- Null Hypothesis: There is no significant difference in the mean stat between Legendary and Non-Legendary Pokémon.
- Alternative Hypothesis: There is a significant difference in the mean stat between Legendary and Non-Legendary Pokémon.

In [16]:
# Load the dataset
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv"
df = pd.read_csv(url)

# Filter the data into Legendary and Non-Legendary groups
legendary = df[df['Legendary'] == True]
non_legendary = df[df['Legendary'] == False]

# List of stats to compare
stats_to_compare = ['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']

# Perform independent t-tests for each stat
results = {}
for stat in stats_to_compare:
    t_stat, p_value = ttest_ind(legendary[stat], non_legendary[stat], equal_var=False)
    results[stat] = {'T-statistic': t_stat, 'P-value': p_value}

# Output the results
for stat, result in results.items():
    print(f"{stat}: T-statistic = {result['T-statistic']:.4f}, P-value = {result['P-value']:.4f}")
    
# Apply Bonferroni correction for multiple comparisons (optional)
alpha = 0.05 / len(stats_to_compare)
print(f"\nBonferroni-corrected significance level: {alpha:.4f}")

HP: T-statistic = 8.9814, P-value = 0.0000
Attack: T-statistic = 10.4381, P-value = 0.0000
Defense: T-statistic = 7.6371, P-value = 0.0000
Sp. Atk: T-statistic = 13.4174, P-value = 0.0000
Sp. Def: T-statistic = 10.0157, P-value = 0.0000
Speed: T-statistic = 11.4750, P-value = 0.0000

Bonferroni-corrected significance level: 0.0083


__Interpretation of Results:__
The t-tests were performed for six stats (HP, Attack, Defense, Sp. Atk, Sp. Def, Speed), comparing Legendary and Non-Legendary Pokémon. Here are the key results:

- T-statistics and p-values for all stats are extremely low (all are 0.0000), meaning the differences are highly significant.
- The Bonferroni-corrected significance level is 0.0083 (since we divided 0.05 by 6, the number of tests).

__Conclusion:__
Since all the p-values are well below the corrected threshold (0.0083), we can conclude the following:

- There is strong evidence to reject the null hypothesis for all six stats.
- This means that there are significant differences between Legendary and Non-Legendary Pokémon for all six stats (HP, Attack, Defense, Special Attack, Special Defense, and Speed).
In other words, Legendary Pokémon tend to have significantly higher stats across the board compared to Non-Legendary Pokémon.

<br>
<br>
<br>

**Challenge 2**

In this challenge, we will be working with california-housing data. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv

In [27]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv")
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -114.31     34.19                15.0       5612.0          1283.0   
1    -114.47     34.40                19.0       7650.0          1901.0   
2    -114.56     33.69                17.0        720.0           174.0   
3    -114.57     33.64                14.0       1501.0           337.0   
4    -114.57     33.57                20.0       1454.0           326.0   

   population  households  median_income  median_house_value  
0      1015.0       472.0         1.4936             66900.0  
1      1129.0       463.0         1.8200             80100.0  
2       333.0       117.0         1.6509             85700.0  
3       515.0       226.0         3.1917             73400.0  
4       624.0       262.0         1.9250             65500.0

**We posit that houses close to either a school or a hospital are more expensive.**

- School coordinates (-118, 37)
- Hospital coordinates (-122, 34)

We consider a house (neighborhood) to be close to a school or hospital if the distance is lower than 0.50.

Hint:
- Write a function to calculate euclidean distance from each house (neighborhood) to the school and to the hospital.
- Divide your dataset into houses close and far from either a hospital or school.
- Choose the propper test and, with 5% significance, comment your findings.
 

In [29]:
# Coordinates of school and hospital
school_coords = (-118, 37)
hospital_coords = (-122, 34)

# Function to calculate Euclidean distance
def calculate_distance(lon, lat, ref_lon, ref_lat):
    return np.sqrt((lon - ref_lon)**2 + (lat - ref_lat)**2)

# Calculate distances to the school and hospital
df['distance_to_school'] = calculate_distance(df['longitude'], df['latitude'], *school_coords)
df['distance_to_hospital'] = calculate_distance(df['longitude'], df['latitude'], *hospital_coords)

# Classify houses as "close" if distance is less than 0.50 to either school or hospital
df['close_to_school_or_hospital'] = (df['distance_to_school'] < 0.50) | (df['distance_to_hospital'] < 0.50)

# Split the dataset into houses close and far from either a school or hospital
close_houses = df[df['close_to_school_or_hospital'] == True]
far_houses = df[df['close_to_school_or_hospital'] == False]

# Perform independent t-test on house prices (median_house_value)
t_stat, p_value = ttest_ind(close_houses['median_house_value'], far_houses['median_house_value'], equal_var=False)

# Output the results
print(f"T-statistic: {t_stat:.4f}, P-value: {p_value:.4f}")

# Significance level
alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis: Houses close to a school or hospital are significantly more expensive.")
else:
    print("Fail to reject the null hypothesis: No significant difference in house prices.")

T-statistic: -17.1742, P-value: 0.0001
Reject the null hypothesis: Houses close to a school or hospital are significantly more expensive.


Interpretation of the Results:
T-statistic: -17.1742
P-value: 0.0001
Since the p-value (0.0001) is much smaller than 0.05, we can reject the null hypothesis. This means there is strong statistical evidence to support the claim that houses close to a school or hospital are, on average, significantly more expensive compared to houses that are farther away.